### ISSUE TITLE
FTP: 500 'LIST ': command not understood

### ISSUE THREAD

**Issue by sschaetz** [rclone/rclone#1850]
I'm using `rclone-v1.38-175-g159fce01β-osx-amd64`
trying to talk to `FTP (Solaris ftpd)`
running `clone ls my_ftp: -vvv`

I get:

2017/11/22 21:51:34 DEBUG : Using config file from "/Users/me/.config/rclone/rclone.conf"
2017/11/22 21:51:34 DEBUG : rclone: Version "v1.38-175-g159fce01β" starting with parameters ["rclone" "ls" "my_ftp:" "-vvv"]
2017/11/22 21:51:34 DEBUG : ftp://abc:21: Connecting to FTP server
2017/11/22 21:51:36 INFO  : ftp://abc:21: Modify window not supported
2017/11/22 21:51:37 ERROR : : error listing: 500 'LIST ': command not understood.
2017/11/22 21:51:37 Failed to ls: 500 'LIST ': command not understood.

**Comment by riopardo** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  monitoramento@wktecno.com
    Domain wktecnomedia.com has exceeded the max emails per hour (362/300 (120%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o8.sgmail.github.com ([167.89.101.199]:4439)
	by hcomm35.dns-wk.info with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89)
	(envelope-from <bounces+848413-9a47-wao=agencia.pw@sgmail.github.com>)
	id 1eHc3c-0001ZT-Cr
	for wao@agencia.pw; Wed, 22 Nov 2017 18:55:57 -0200
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=eGH3qEmCm2NhKQwD5rySS6QNg/0=; b=uB1XqaUhAI2vMgb8
	56KEGRhnVBg3VQ6u0Q3AchROORmwD8ER4b+v2Ok3vWl7RKY+si5gdhIWRJNVx7w5
	zearSCJPX+txmuCWpsC4otW2Bxd2FVEhGV2kjvHzcgNDAh8+iNefpQa9VmlyKNaY
	A+gVC79YMcduMPz+56gZdnlia0I=
Received: by filter0215p1iad2.sendgrid.net with SMTP id filter0215p1iad2-13582-5A15E44B-16
        2017-11-22 20:55:39.750332788 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0002p1iad2.sendgrid.net (SG) with ESMTP id R0npYMnIRZWLCZdjXB9saA
	for <wao@agencia.pw>; Wed, 22 Nov 2017 20:55:39.675 +0000 (UTC)
Date: Wed, 22 Nov 2017 20:55:39 +0000 (UTC)

**Comment by ncw** [comment]
I wonder if it is complaining about the space in `LIST `.  All FTP servers should understand `LIST`!

Any chance you could connect to the server with telnet and try `LIST` and `LIST ` and see what happens.  Here is [a guide](https://www.webdigi.co.uk/blog/2009/ftp-using-raw-commands-and-telnet/) if you haven't done that kind of thing before.

**Comment by sschaetz** [comment]
With or without whitespace, both are not supported by this particular FTP server.

Using Cyberduck (which talks to this particular FTP server successfully) I captured a transaction - it does not use `LIST` at all it uses `STAT` only it seems; I could replicate that manually through telnet too. Suspicion is that this is some fairly old or esoteric FTP server software I'm dealing with.

**Comment by ncw** [comment]
Hmm interesting... LIST is defined by the [original FTP RFC959](https://tools.ietf.org/html/rfc959)

It looks like the [FTP library](https://github.com/jlaffaye/ftp/issues) I'm using only supports LIST or MLSD, not STAT so I think you are out of luck unless you want to get hacking :-(

**Comment by riopardo** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  monitoramento@wktecno.com
    Domain wktecnomedia.com has exceeded the max emails per hour (349/300 (116%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o3.sgmail.github.com ([192.254.112.98]:62700)
	by hcomm35.dns-wk.info with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89)
	(envelope-from <bounces+848413-9a47-wao=agencia.pw@sgmail.github.com>)
	id 1eHoY4-0007xp-Mf
	for wao@agencia.pw; Thu, 23 Nov 2017 08:16:15 -0200
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=Mz5S0z3CIirJJYFwbyy6IysywLs=; b=I5nv9K236UGMDi2k
	Tl7f3sCpbTwmaTvV8Yvpi1fIV16hLN7KEpt+iyafWnY5FpaAQSITNQ9svjtxT4nY
	VhtoA9JpJAqEbnuNJ5pzZkrpYquGrc3HnBestSqzQ3KbDtGBMEJxCPdXnDaGNpSj
	6OQH8O/sKTRI+LTI33QR5dm0ax4=
Received: by filter0488p1las1.sendgrid.net with SMTP id filter0488p1las1-11350-5A169FDD-1
        2017-11-23 10:15:57.08812007 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0003p1iad1.sendgrid.net (SG) with ESMTP id WzQOvjebRdSa_fpVv98k7A
	for <wao@agencia.pw>; Thu, 23 Nov 2017 10:15:57.015 +0000 (UTC)
Date: Thu, 23 Nov 2017 10:15:57 +0000 (UTC)

**Comment by sschaetz** [comment]
My solution for now is to wget the FTP contents (which supports this FTP behaviour) and then use rclone to pump the data to the final destination.

**Comment by Sash0k** [comment]
got the same with Solaris, probably it can be helpful
send **LIST** 

```
220 ... FTP server ready.
331 Password required for ....
230 User ... logged in.
200 Type set to A.
227 Entering Passive Mode (...)
500 'LIST ': command not understood.
```

send **LIST .** // with dot to use current dir
```
220 ... FTP server ready.
331 Password required for ...
230 User ... logged in.
200 Type set to A.
227 Entering Passive Mode (...)
150 Opening ASCII mode data connection for /bin/ls.
total 80
-rw-------   1 user group           68 file1
drwxr-xr-x   3 user group          512 dir 1
drwxr-xr-x   3 user group          512 dir 2
226 Transfer complete.
221-You have transferred 0 bytes in 0 files.
221-Total traffic for this session was 680 bytes in 1 transfers.
221-Thank you for using the FTP service on ...
221 Goodbye.
```

**Comment by ivandeex** [comment]
@sschaetz Does the problem reproduce with rclone 1.54?

**Comment by sschaetz** [comment]
@ivandeex unfortunately I have no way to test it as I don't have access to that FTP server any longer, sorry.

**Comment by sschaetz** [comment]
Maybe @Sash0k still has access and can test it out?

**Comment by ivandeex** [comment]
@sschaetz @Sash0k ok
I am going to close this with "can't reproduce"
Feel free to resubmit if you have new info.

**Comment by Kyle-Tweedie** [comment]
Just tested it. It's still occurring.

**Comment by ivandeex** [comment]
@Kyle-Tweedie got you

**Comment by ivandeex** [comment]
@Kyle-Tweedie 
Please run `rclone -vv --log-file bug.log REST_OF_YOUR_COMMAND` and attach below. Then extract config section from rclo e.conf, redact password and attach too. Thank you.

**Comment by ncw** [comment]
Adding `--dump bodies` will be useful too - thanks.

**Comment by Kyle-Tweedie** [comment]
@ivandeex @ncw I noticed that the command would work if I added "/" at the end of the source dir. 
So instead of "abcFTP:" I'd use "abcFTP:/" and the latter would work.

[bug.log](https://github.com/rclone/rclone/files/5973939/bug.log)

**Comment by ivandeex** [comment]
Aha! Nice catch!
But the `remote:` (without `/`) form is widespread in rclone. I would think we could internally replace empty path by a single slash to mitigate this for _your_ ftp server. I'm just afraid not it might confuse _other_ ftp servers.

**Comment by Kyle-Tweedie** [comment]
Would it not be possible to include an option in the config under advanced options?

**Comment by ivandeex** [comment]
If it depends on ftp server (breaks some), I'd rather document it as a caveat. People would have to dig docs for that option anyway (then edit rclone.conf, then rerun a command, all for a single slash). If the above workaround helps some servers without breaking others, I'd go for it.

**Comment by ncw** [comment]
In general `ftp:/dir` means look at the directory `/dir` whereas `ftp:dir` means look at the directory `dir` in your home directory.

So they are two different paths.

It does say this in [the docs](https://rclone.org/ftp/)

> Paths are specified as remote:path. If the path does not begin with a / it is relative to the home directory of the user. An empty path remote: refers to the user's home directory.

**Comment by ivandeex** [comment]
> In general `ftp:/dir` means look at the directory `/dir` whereas `ftp:dir` means look at the directory `dir` in your home directory.

Ah, FTP (and I guess SFTP too) is not like storage provider backends I'm used to. Please excuse me for incorrect idea.

Then I guess a `ftpremote:.` (with a dot) could help?

**Comment by Kyle-Tweedie** [comment]
I see. Well a workaround is to create a new remote and select `alias` as the type and then just select the ftp and add a `/` like so `ftpremote:/`. Working for me now in the CMD and web GUI.

**Comment by ivandeex** [comment]
rclone's `ls` with empty on-remote path calls `ftpConnection.List("")` in rclone code https://github.com/rclone/rclone/blob/v1.54.0/backend/ftp/ftp.go#L520, which in turn emits `LIST ` with trailing space in external code https://github.com/jlaffaye/ftp/blob/master/ftp.go#L573
I think the `ftpremote:.` trick will result in the `LIST .` command. If it helped and the problem goes and the new log has something like `LIST . -> 200 OK -> .....`, we would conclude to meet an ftp server with "naughty" command parser, propose immediate workaround and think of next steps (pray upstream to trim trailing space, sacrifice a virgin, replace "" by "." from the scratch, replace after `500`, etc).

**Comment by ncw** [comment]
> pray upstream to trim trailing space,

I've sent lots of patches to jlaffaye/ftp and they've been accepted so if someone wants to have a go at that - it will probably be accepted.

**Comment by ivandeex** [comment]
@Kyle-Tweedie 
Could you please disable workarounds for a moment and try this rclone beta instead?
https://github.com/ivandeex/rclone/releases/tag/v1.54-ftp01
Does it solve the problem for you?
(just for reference, the beta fix is https://github.com/ivandeex/ftp/commit/7c9cab6372b53639bb3f29f86e994fc40668ac7a)
Hope it helps!
Ivan

**Comment by Kyle-Tweedie** [comment]
@ivandeex That fixes it! Working perfectly now.

**Comment by ivandeex** [comment]
Great!
Having your confirmation I am now going to submit a change to upstream, wait for @jlaffaye reaction and integrate changes back if all goes well. It can take a while before we see the fix in an official rclone release.
Thank you

**Comment by ivandeex** [comment]
Pending on https://github.com/jlaffaye/ftp/pull/219

**Comment by ivandeex** [comment]
@Kyle-Tweedie 

We've integrated the fix.
Could you please retest with the latest beta https://beta.rclone.org/v1.55.0-beta.5243.75056dc9b/ ?

**Comment by charlesvestal** [comment]
Having previously had this issue on a FRITZ!Box 7530, I can confirm the beta above resolves the 500 and functions as expected without a trailing slash.
